addons/sale_coupon/tests/test_sale_invoicing.py

```diff
@@ -50,3 +50,94 @@ def test_invoicing_order_with_promotions(self):
         self.assertEqual(order.order_line, invoiceable_lines)
         account_move = order._create_invoices()
         self.assertEqual(len(account_move.invoice_line_ids), 2)
+
+    def test_double_invoice_order_with_order_percentage_discount(self):
+        discount_coupon_program = self.env['sale.coupon.program'].create({
+            'name': '10% Discount',
+            'program_type': 'coupon_program',
+            'reward_type': 'discount',
+            'discount_apply_on': 'on_order',
+            'promo_code_usage': 'no_code_needed',
+        })
+        product = self.env['product.product'].create({
+            'invoice_policy': 'order',
+            'name': 'Product invoiced on order',
+            'lst_price': 800,
+        })
+
+        order = self.empty_order
+        order.write({
+            'order_line': [
+                (0, 0, {
+                    'product_id': product.id,
+                })
+            ]
+        })
+
+        order.recompute_coupon_lines()
+        order.action_confirm()
+
+        account_move = order._create_invoices()
+        self.assertEqual(len(account_move.invoice_line_ids), 2)
+        self.assertEqual(account_move.amount_untaxed, (product.list_price - discount_coupon_program.discount_percentage * product.list_price / 100.0))
+
+        order.write({
+            'order_line': [
+                (0, 0, {
+                    'product_id': product.id,
+                })
+            ]
+        })
+
+        order.recompute_coupon_lines()   
+
+        second_account_move = order._create_invoices()
+        self.assertEqual(len(second_account_move.invoice_line_ids), 2)
+        self.assertEqual(second_account_move.amount_untaxed, (product.list_price - discount_coupon_program.discount_percentage * product.list_price / 100.0))
+
+    def test_double_invoice_order_with_free_product(self):
+        order = self.empty_order
+        product = self.env['product.product'].create({
+            'invoice_policy': 'order',
+            'name': 'Product invoiced on order',
+            'lst_price': 800,
+        })
+        self.env['sale.coupon.program'].create({
+            'name': 'Buy 4 Large Desks, one is free',
+            'reward_type': 'product',
+            'reward_product_id': product.id,
+            'rule_products_domain': "[('id', 'in', [%s])]" % (product.id),
+            'discount_apply_on': 'on_order',
+            'promo_code_usage': 'no_code_needed',
+            'rule_min_quantity': 3.0
+        })
+        order.write({
+            'order_line': [
+                (0, 0, {
+                    'product_id': product.id,
+                    'product_uom_qty': 4.0,
+                })
+            ]
+        })
+
+        order.recompute_coupon_lines()
+        order.action_confirm()
+
+        account_move = order._create_invoices()
+        self.assertEqual(len(account_move.invoice_line_ids), 2)
```

Comment:
Again personal opinion, but I'd always try to add the third parameter message when writing an `assertXXX()`, especially in coupon tests (or anything related to numbers), indeed:
1. It is easier to grep once you ran into the error and have to find it inside your IDE
2. It is easier for the one reading it later
Eg
```python
self.assertEqual(len(order.order_line.ids), 2, "We should have 2 lines as we now have one 'Free Large Cabinet' line as we bought 4 of them")
```
is easier tu figure than
```python
self.assertEqual(len(order.order_line.ids), 2)
```

As previous comment, ignore if you don't really care about that